### PR TITLE
Remove ContentTypes dependency

### PR DESCRIPTION
**Pull request by lynnco**
closes https://github.com/SexualHealthInnovations/django-wizard-builder/issues/39

- [x] remove django-polymorphic from PageBase
- [x] remove django-polymorphic from FormQuestion
- [x] add InheritanceManager to PageBase
- [x] add InheritanceManager to FormQuestion
- [x] check fk to QuestionPage (not PageBase!) on FormQuestion
- [x] check fk to MultipleChoice (not FormQuestion!) on Choice
- [x] check fk to PageBase on Conditional
- [x] check fk to FormQuestion on Conditional
- [x] do all of this without changing the current tests
- [x] update admin to use down-casting
- [x] check fk to FormQuestion on callisto-core.evaluation.EvaluationField (in https://github.com/SexualHealthInnovations/callisto-core) (won't be in this PR)

### CODE REVIEW

**Comment by lynnco** [comment]
WIP !

**Comment by lynnco** [comment]
It looks like downcasting on related objects happens automatically? Or maybe it knows to check for QuestionPage? https://github.com/SexualHealthInnovations/django-wizard-builder/pull/40/files#diff-ba77455cb828887e1eb42cdfd40bc43d

**Comment by lynnco** [comment]
ah here we go

**Comment by lynnco** [comment]
lmao python 3.2 is failing in codecov??? ok, removing it from this PR then

**Comment by lynnco** [comment]
Alright! admin and evaluation are the last bits

File: .travis.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ matrix:
         - python: "3.3"
           env: TOXENV=django110
 install:
-    - pip install tox codecov isort flake8
+    - pip install -r requirements-test.txt
```
Comment (lynnco) [inline .travis.yml]: cleanup. I'm honestly not 100% sure if / why this was needed

File: tox.ini [hunk 2]
```diff
@@ -11,4 +11,4 @@ commands = coverage run runtests.py
 deps =
     django18: Django>=1.8,<1.9
     django110: Django>=1.10,<1.11
-    -r{toxinidir}/requirements-test.txt
+    -rrequirements-test.txt
```
Comment (lynnco) [inline tox.ini]: more cleanup! python 3.2 wasnt actually being tested on travis, and for whatever reason the requirements weren't updating

**Comment by lynnco** [comment]
✔️ admin

File: tests/test_inheritence.py [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        self.assertIsInstance(choice.question, MultipleChoice)
+
+    def test_conditional_instance(self):
+        page = QuestionPage.objects.create()
+        question = MultipleChoice.objects.create(page_id=page.id)
+        Conditional.objects.create(question_id=question.id, page_id=page.id)
+        condition = Conditional.objects.first()
+        self.assertTrue(QuestionPage.objects.first().conditional)
+        self.assertIsInstance(condition.question, MultipleChoice)
+        self.assertIsInstance(condition.page, QuestionPage)
```
Comment (lynnco) [inline tests/test_inheritence.py]: attribute backwards compatibility tests

File: wizard_builder/admin/base.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+from django.contrib import admin
+from django.core.urlresolvers import reverse_lazy
+from django.db import models
+from django.urls import reverse
+from django.utils.html import format_html
+
+from ..models import Choice, FormQuestion
+
+
+class DowncastedAdmin(admin.ModelAdmin):
```
Comment (lynnco) [inline wizard_builder/admin/base.py]: this is for replicating a bunch of the polymorphic admin functionality

File: wizard_builder/admin/page_admin.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from django.contrib import admin
+
+from .base import DowncastedAdmin, QuestionInline
+
+
+class PageBaseAdmin(DowncastedAdmin):
+    list_filter = ['site']
```
Comment (lynnco) [inline wizard_builder/admin/page_admin.py]: added site awareness to admin in this PR, since I was here

File: wizard_builder/admin/question_admin.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+from .base import ChoiceInline, DowncastedAdmin
+
+
+class FormQuestionAdminMixin(object):
+    search_fields = ['text', 'descriptive_text']
+    list_filter = ['page__site']
+
+
+class FormQuestionParentAdmin(FormQuestionAdminMixin, DowncastedAdmin):
+    list_display = ['short_str', 'model_type', 'page']
```
Comment (lynnco) [inline wizard_builder/admin/question_admin.py]: `short_str` was created because I made `__str__` a ton more verbose

File: wizard_builder/migrations/0004_inheritence_downcasting.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+        ),
+        migrations.RemoveField(
+            model_name='formquestion',
+            name='polymorphic_ctype',
+        ),
+        migrations.RemoveField(
+            model_name='pagebase',
+            name='polymorphic_ctype',
+        ),
+    ]
```
Comment (lynnco) [inline wizard_builder/migrations/0004_inheritence_downcasting.py]: this migration is entirely deletions :DDDDD

File: wizard_builder/models.py [hunk 8]
```diff
@@ -28,6 +26,17 @@ class PageBase(PolymorphicModel):
     site = models.ForeignKey(Site, null=True)
     objects = PageBaseManager()
 
+    @property
+    def short_str(self):
```
Comment (lynnco) [inline wizard_builder/models.py]: used by admin

**Comment by lynnco** [comment]
wish I knew what was going on with the requirements :/

File: tests/test_inheritence.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
+from wizard_builder.models import (
+    Choice, Conditional, FormQuestion, MultipleChoice, PageBase, QuestionPage,
+    SingleLineText,
+)
+
+from django.test import TestCase
+
+
+class InheritenceTest(TestCase):
```
Comment (kelseyq) [inline tests/test_inheritence.py]: inheritAnce (typo)

File: tox.ini (hunk 2 above)
Comment (kelseyq) [inline tox.ini]: does it need a space after the `-r`?

Comment (lynnco) [inline tox.ini]: no https://stackoverflow.com/questions/12258456/how-can-tox-install-the-modules-via-the-requirements-file#comment38660198_18964795 dunno why though

**Comment by lynnco** [comment]
codecov is giving a 👎  because there's no test coverage for admin

and like, that's super reasonable IMO. #AdminBlindspot

**Comment by kelseyq** [comment]
nooooo not a hashtag!!!

**Comment by kelseyq** [comment]
can you write out what manual testing you did? maybe in a (sigh) #adminblindspot ticket?

**Comment by lynnco** [comment]
I think I can actually write an integration test to cover this, I just finished writing essentially the same test for callisto 🔒 . Changing it to test wizard builder instead would take me maybe 10 minutes

File: wizard_builder/migrations/0004_inheritence_downcasting.py (hunk 7 above)
Comment (kelseyq) [inline wizard_builder/migrations/0004_inheritence_downcasting.py]: so IDs don't change, right? can you confirm that it's possible to successfully edit a record created pre-migration after this migration runs?

Comment (lynnco) [inline wizard_builder/migrations/0004_inheritence_downcasting.py]: id's definitely don't change, but I do definitely want to add a test for editing

**Comment by lynnco** [comment]
I really need to get tox working on my local setup :/

**Comment by lynnco** [comment]
(partial fail, only django 1.10 working so far)

File: .travis.yml [hunk 10]
```diff
@@ -5,9 +5,6 @@ python:
     - "3.4"
     - "3.5"
 sudo: false
-cache:
-    directories:
-        - $TRAVIS_BUILD_DIR/.tox
```
Comment (lynnco) [inline .travis.yml]: travis was caching the dependency installs :/

File: tests/settings.py [hunk 11]
```diff
@@ -10,24 +8,48 @@
...
-    "django.contrib.auth",
-    "django.contrib.contenttypes",
-    "django.contrib.sessions",
-    "django.contrib.sites",
+    'django.contrib.admin',
+    'django.contrib.auth',
+    'django.contrib.sites',
+    'django.contrib.contenttypes',
+    'django.contrib.sessions',
+    'django.contrib.messages',
```
Comment (kelseyq) [inline tests/settings.py]: what requires these now?

File: tests/settings.py [hunk 12]
```diff
@@ -10,24 +8,48 @@
...
+    'django.contrib.staticfiles',
     "wizard_builder",
     "tests.test_app"
 ]
 
-MIDDLEWARE_CLASSES = ('django.contrib.sessions.middleware.SessionMiddleware',)
+MIDDLEWARE_CLASSES = (
+    'django.contrib.sessions.middleware.SessionMiddleware',
```
Comment (kelseyq) [inline tests/settings.py]: can you summarize the middleware & templates changes for me?

Comment (kelseyq) [inline tests/settings.py]: ohhh is all of this for functional tests?

Comment (kelseyq) [inline tests/settings.py]: whoops i didn't look at the filename! neeeevvver mind

Comment (lynnco) [inline tests/settings.py]: yep, was just typing that

File: tests/test_admin.py [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+
+
+User = get_user_model()
+
+SCREEN_DUMP_LOCATION = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)), 'screendumps'
+)
+
+
+class FunctionalTest(StaticLiveServerTestCase):
```
Comment (kelseyq) [inline tests/test_admin.py]: i bet getting all this working in the diff python & django versions was fun!

Comment (lynnco) [inline tests/test_admin.py]: 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭 😭

Comment (lynnco) [inline tests/test_admin.py]: I mean, ok, in working on that I got the entire tox build matrix passing on local, which was kind of cool I guess
